[JSC] Fix spec compliance and optimize `RegExp#@@search`

sosukesuzuki opened the pull request:
#### 8657eef539f38caf5a8066aeeac5524832a35bed
<pre>
[JSC] Fix spec compliance and optimize `RegExp#@@search`
<a href="https://bugs.webkit.org/show_bug.cgi?id=288832">https://bugs.webkit.org/show_bug.cgi?id=288832</a>

Reviewed by Yusuke Suzuki.

According to the spec[1], `RegExp#[Symbol.search]` writes to the `lastIndex`
property of `this` in the following cases:

- `lastIndex` is not 0
- The value of `lastIndex` changes before and after calling `RegExpExec`
  (which occurs when the `g` or `y` flag is enabled)

So, if `lastIndex` is not writable and the above conditions are met, a
`TypeError` should be thrown.

However, in the current JSC implementation, no error is thrown. So we
should change to execute the slow path when `lastIndex` is not writable,
ensuring that an error is thrown.

Furthermore, adding additional fast-path checks to the current JavaScript
implementation might lead to a performance regression[2].

Thus, this patch not only fixes the spec compliance issue but also implements
the behavior in C++ and applies optimizations in DFG/FTL.

Below are the microbenchmark results when DFG/FTL is enabled:

                                                  TipOfTree                  Patched

regexp-prototype-search-short-string           80.3555+-0.4927     ^     76.5456+-0.6090        ^ definitely 1.0498x faster
regexp-prototype-search-complex-pattern        93.2712+-0.3697     ^     89.3464+-0.7167        ^ definitely 1.0439x faster
regexp-prototype-search-basic                 123.5857+-3.7591          120.8217+-0.4638          might be 1.0229x faster
regexp-prototype-search-anchor                100.5400+-1.0777     ^     96.3188+-1.4436        ^ definitely 1.0438x faster

[1]: <a href="https://tc39.es/ecma262/#sec-regexp.prototype-%symbol.search%">https://tc39.es/ecma262/#sec-regexp.prototype-%symbol.search%</a>
[2]: <a href="https://github.com/WebKit/WebKit/pull/41631#discussion_r1982656344">https://github.com/WebKit/WebKit/pull/41631#discussion_r1982656344</a>

* JSTests/microbenchmarks/regexp-prototype-search-anchor.js: Added.
(shouldBe):
* JSTests/microbenchmarks/regexp-prototype-search-basic.js: Added.
(shouldBe):
* JSTests/microbenchmarks/regexp-prototype-search-complex-pattern.js: Added.
(shouldBe):
* JSTests/microbenchmarks/regexp-prototype-search-short-string.js: Added.
(shouldBe):
* JSTests/stress/regexp-prototype-search-fastpath-check.js: Added.
(shouldBe):
(shouldThrow):
(i.shouldThrow):
(i.get called):
(get return):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
(JSC::DFG::FixupPhase::addRegExpSearchPrimordialChecks):
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasHeapPrediction):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
(JSC::RegExpPrototype::finishCreation):
(JSC::regExpExec):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/RegExpPrototype.h:
* Source/JavaScriptCore/runtime/RegExpPrototypeInlines.h:
(JSC::regExpExecWatchpointIsValid):
(JSC::regExpTestWatchpointIsValid): Deleted.

Canonical link: <a href="https://commits.webkit.org/296443@main">https://commits.webkit.org/296443@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/903b836a95a7a34fbd0ece3af82f05a2fa8c2c69

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108552 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28213 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18637 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113761 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58951 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36767 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82446 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111500 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22934 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97776 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62882 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22351 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15916 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58476 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101109 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92304 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15967 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/rotated-cat-off-top-edge.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116882 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107111 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35606 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26249 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91472 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35979 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94048 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91273 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23251 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36165 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13931 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31369 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35508 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41043 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131393 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35220 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35645 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38569 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36898 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -1675,6 +1676,49 @@ JSC_DEFINE_JIT_OPERATION(operationRegExpTestGeneric, size_t, (JSGlobalObject* gl
+    JSString* string = argument.toStringOrNull(globalObject);
+    EXCEPTION_ASSERT(!!scope.exception() == !string);
+    if (!string)
+        OPERATION_RETURN(scope, false);
```
`false` looks not correct for the result using `UCPUStrictInt32`.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -13345,6 +13345,56 @@ void SpeculativeJIT::compileRegExpMatchFast(Node* node)
+            TrustedImm32(RegExpObject::lastIndexIsNotWritableFlag)));
+
+    flushRegisters();
+    GPRFlushedCallResult result(this);
```
Let's define `resultGPR = ...`;

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -16972,6 +16975,25 @@ IGNORE_CLANG_WARNINGS_END
+                m_out.constIntPtr(RegExpObject::lastIndexIsNotWritableFlag)));
+        if (m_node->child3().useKind() == StringUse) {
+            LValue argument = lowString(m_node->child3());
+            setInt32(vmCall(Int32, operationRegExpSearch, globalObject, base, argument));
```
Isn't it `operationRegExpSearchString`?

Constellation reviewed (changes requested):
Looks good direction!

Constellation commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -432,22 +450,61 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoGetterSource, (JSGlobalObject* globalObject,
+
+    auto resultIndex = match.get(globalObject, vm.propertyNames->index);
+    RETURN_IF_EXCEPTION(scope, { });
+    return JSValue::encode(resultIndex);
```
Do,
```
RELEASE_AND_RETURN(scope, JSValue::encode(match.get(globalObject, vm.propertyNames->index));
```

Constellation commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -432,22 +450,61 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoGetterSource, (JSGlobalObject* globalObject,
+    return JSValue::encode(resultIndex);
 }
 
+
```
Not necessary line.

Constellation commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -432,22 +450,61 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoGetterSource, (JSGlobalObject* globalObject,
+            RETURN_IF_EXCEPTION(scope, { });
+            MatchResult result = globalObject->regExpGlobalData().performMatch(globalObject, regExp->regExp(), str, strView, 0);
+            RETURN_IF_EXCEPTION(scope, { });
+            return JSValue::encode(result ? jsNumber(result.start) : jsNumber(-1));
```
You can do,
```
scope.release();
MatchResult result = globalObject->regExpGlobalData().performMatch(globalObject, regExp->regExp(), str, strView, 0);
return JSValue::encode(result ? jsNumber(result.start) : jsNumber(-1));
```

Constellation commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -432,22 +450,61 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoGetterSource, (JSGlobalObject* globalObject,
+
+    auto currentLastIndex = thisObject->get(globalObject, vm.propertyNames->lastIndex);
+    RETURN_IF_EXCEPTION(scope, { });
+    if (!sameValue(globalObject, currentLastIndex, previsouLastIndex)) {
```
You need exception check after `sameValue` call.
So,
```
bool isSameValue = sameValue(globalObject, currentLastIndex, previsouLastIndex);
RETURN_IF_EXCEPTION(scope, { });
if (!isSameValue)
```

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -1675,6 +1676,49 @@ JSC_DEFINE_JIT_OPERATION(operationRegExpTestGeneric, size_t, (JSGlobalObject* gl
+    OPERATION_RETURN_IF_EXCEPTION(scope, 0);
+
+    MatchResult result = globalObject->regExpGlobalData().performMatch(globalObject, regExpObject->regExp(), string, strView, 0);
+    OPERATION_RETURN_IF_EXCEPTION(scope, 0);
```
Remove this check since `OPERATION_RETURN` will check it.

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -1675,6 +1676,49 @@ JSC_DEFINE_JIT_OPERATION(operationRegExpTestGeneric, size_t, (JSGlobalObject* gl
+    OPERATION_RETURN_IF_EXCEPTION(scope, 0);
+
+    MatchResult result = globalObject->regExpGlobalData().performMatch(globalObject, regExpObject->regExp(), string, strView, 0);
+    OPERATION_RETURN_IF_EXCEPTION(scope, 0);
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -13408,6 +13408,57 @@ void SpeculativeJIT::compileRegExpMatchFast(Node* node)
+    GPRReg globalObjectGPR = globalObject.gpr();
+
+    SpeculateCellOperand base(this, node->child2());
+    GPRReg baseGPR = base.gpr();
```
We would like to do,
```
SpeculateCellOperand globalObject(this, node->child1());
SpeculateCellOperand base(this, node->child2());

GPRReg globalObjectGPR = globalObject.gpr();
GPRReg baseGPR = base.gpr();
```

Constellation commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -1719,6 +1719,18 @@ class FixupPhase : public Phase {
+            if (m_graph.isWatchingRegExpPrimordialPropertiesWatchpoint(node))
+                addRegExpSearchPrimordialChecks(node->child2().node());
+            else
+                m_insertionSet.insertNode(m_indexInBlock, SpecNone, ForceOSRExit, node->origin);
```
This is not correct. If we emit ForceOSRExit here, we should have a code which does not use `RegExpSearch` when watchpoint is already fired. But DFGByteCodeParser is not checking that status. So we will have OSR exit loop.

sosukesuzuki replied:
Added `regExpPrimordialPropertiesWatchpointSet().isStillValid()` checking to `DFGBytecodeParser`

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -3194,6 +3194,57 @@ auto ByteCodeParser::handleIntrinsicCall(Node* callee, Operand resultOperand, Ca
+                return CallOptimizationResult::DidNothing;
+
+            // Don't inline intrinsic if we exited due to one of the primordial RegExp checks failing.
+            if (m_inlineStackTop->m_exitProfile.hasExitSite(m_currentIndex, BadConstantValue))
```
We should have ExoticObject exit guard too

sosukesuzuki committed:
[JSC] Fix spec compliance and optimize `RegExp#@@search`
https://bugs.webkit.org/show_bug.cgi?id=288832

Reviewed by Yusuke Suzuki.

According to the spec[1], `RegExp#[Symbol.search]` writes to the `lastIndex`
property of `this` in the following cases:

- `lastIndex` is not 0
- The value of `lastIndex` changes before and after calling `RegExpExec`
  (which occurs when the `g` or `y` flag is enabled)

So, if `lastIndex` is not writable and the above conditions are met, a
`TypeError` should be thrown.

However, in the current JSC implementation, no error is thrown. So we
should change to execute the slow path when `lastIndex` is not writable,
ensuring that an error is thrown.

Furthermore, adding additional fast-path checks to the current JavaScript
implementation might lead to a performance regression[2].

Thus, this patch not only fixes the spec compliance issue but also implements
the behavior in C++ and applies optimizations in DFG/FTL.

Below are the microbenchmark results when DFG/FTL is enabled:

                                                  TipOfTree                  Patched

regexp-prototype-search-short-string           80.3555+-0.4927     ^     76.5456+-0.6090        ^ definitely 1.0498x faster
regexp-prototype-search-complex-pattern        93.2712+-0.3697     ^     89.3464+-0.7167        ^ definitely 1.0439x faster
regexp-prototype-search-basic                 123.5857+-3.7591          120.8217+-0.4638          might be 1.0229x faster
regexp-prototype-search-anchor                100.5400+-1.0777     ^     96.3188+-1.4436        ^ definitely 1.0438x faster

[1]: https://tc39.es/ecma262/#sec-regexp.prototype-%symbol.search%
[2]: https://github.com/WebKit/WebKit/pull/41631#discussion_r1982656344

* JSTests/microbenchmarks/regexp-prototype-search-anchor.js: Added.
(shouldBe):
* JSTests/microbenchmarks/regexp-prototype-search-basic.js: Added.
(shouldBe):
* JSTests/microbenchmarks/regexp-prototype-search-complex-pattern.js: Added.
(shouldBe):
* JSTests/microbenchmarks/regexp-prototype-search-short-string.js: Added.
(shouldBe):
* JSTests/stress/regexp-prototype-search-fastpath-check.js: Added.
(shouldBe):
(shouldThrow):
(i.shouldThrow):
(i.get called):
(get return):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
(JSC::DFG::FixupPhase::addRegExpSearchPrimordialChecks):
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasHeapPrediction):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
(JSC::RegExpPrototype::finishCreation):
(JSC::regExpExec):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/RegExpPrototype.h:
* Source/JavaScriptCore/runtime/RegExpPrototypeInlines.h:
(JSC::regExpExecWatchpointIsValid):
(JSC::regExpTestWatchpointIsValid): Deleted.

Canonical link: https://commits.webkit.org/296443@main
